### PR TITLE
CB-8761 [wp8]: Copy cookies from WebBrowser

### PR DESCRIPTION
**Pull request by dpolivy**
On other platforms, the cookies from the browser that apply to the
request are automatically included. On WP8, that's not the case.

This commit will copy the cookies from the WebBrowser control, if the
scheme and host of the WebBrowser match that of the HttpWebRequest. For
users who host their web components remotely, and perform
uploads/downloads from the same server using cookie based
authentication, this fix enables that scenario on WP8, and brings it up
to parity with iOS, Android, and others.

Fixes https://issues.apache.org/jira/browse/CB-8761

### CODE REVIEW

**Comment by dpolivy** [comment]
@stevengill Is there someone who can review and merge this?

**Comment by nikhilkh** [comment]
@vladimir-kotikov Can you please take a look?

File: src/wp/FileTransfer.cs [hunk 1]
```diff
@@ -210,6 +216,77 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
...
+                        }
+                    }
+                }
+
+                try
+                {
+                    // Only copy the cookies if the scheme and host match (to avoid any issues with secure/insecure cookies)
+                    // NOTE: since the returned CookieCollection appears to munge the original cookie's domain value in favor of the actual Source domain,
+                    // we can't know for sure whether the cookies would be applicable to any other hosts, so best to play it safe and skip for now.
+                    if (this.browser.Source.Scheme == webRequest.RequestUri.Scheme && this.browser.Source.Host == webRequest.RequestUri.Host)
```
Comment (vladimir-kotikov) [inline src/wp/FileTransfer.cs]: it is very unlikely, but `browser` still can be null here, consider adding check for 'not null' case.

Also `browser.Source.Scheme` and `browser.Source.Host` will throw an `InvalidOperationException` if `Source` is an relative URI. IMO it's better to check if `browser.Source.IsAbsoluteUri == true` than catch an exception here.

File: src/wp/FileTransfer.cs [hunk 2]
```diff
@@ -210,6 +216,77 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
...
+                        string cookieHeader = "";
+                        string requestPath = webRequest.RequestUri.PathAndQuery;
+                        CookieCollection cookies = this.browser.GetCookies();
+
+                        // Iterate over the cookies and add to the header
+                        foreach (Cookie cookie in cookies)
+                        {
+                            // Check that the path is allowed, first
+                            // NOTE: Path always seems to be empty for now, even if the cookie has a path set by the server.
+                            if (cookie.Path.Length == 0 || requestPath.IndexOf(cookie.Path) == 0)
```
Comment (vladimir-kotikov) [inline src/wp/FileTransfer.cs]: `requestPath.IndexOf(cookie.Path)` - this comparison is case-sensitive. Consider using overload of `indexOf` method with `InvariantCultureIgnoreCase` or `OrdinalIgnoreCase` comparison.

File: src/wp/FileTransfer.cs [hunk 3]
```diff
@@ -210,6 +216,77 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
...
+                        if (cookieHeader.Length > 0)
+                        {
+                            webRequest.Headers["Cookie"] = cookieHeader;
+                        }
+                    }
+                    tcs.SetResult(Type.Missing);
+                }
+                catch (Exception ex)
+                {
+                    tcs.SetException(ex);
```
Comment (vladimir-kotikov) [inline src/wp/FileTransfer.cs]: This exception will never be caught. Consider silently return from this method here or add try/catch block around this method call in `upload` and `download`.

**Comment by dpolivy** [comment]
@vladimir-kotikov Thanks for the comments! I addressed your feedback, and pushed up a new diff. I squashed the changes into a single commit -- hope that is OK (I'm not sure what the correct process is for that, but hopefully this makes it an easier merge for you).

**Comment by sgrebnov** [comment]
LGTM

File: src/wp/FileTransfer.cs [hunk 4]
```diff
@@ -210,6 +216,80 @@ public FileTransferProgress(long bTotal = 0, long bLoaded = 0)
...
+        /// </summary>
+        /// <param name="browser">The source browser to copy the cookies from</param>
+        /// <param name="webRequest">The destination HttpWebRequest to add the cookie header to</param>
+        /// <returns>Nothing</returns>
+        private async Task CopyCookiesFromWebBrowser(HttpWebRequest webRequest)
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            // Accessing WebBrowser needs to happen on the UI thread
+            Deployment.Current.Dispatcher.BeginInvoke(() =>
```
Comment (sgrebnov) [inline src/wp/FileTransfer.cs]: I think we can use Invoke() instead of BeginInvoke() here so we don't need to additionally use TaskCompletionSource.

Comment (dpolivy) [inline src/wp/FileTransfer.cs]: @sgrebnov I think the Dispatcher only has the async methods on WP, unfortunately. If you have another suggestion, though, I'm happy to look at it.

Comment (sgrebnov) [inline src/wp/FileTransfer.cs]: @dpolivy I've double checked and you are right, Invoke does not exit in WP8 .Net framework subset.  So please ignore my comment..

**Pull request closed without merging**